src/liboslexec/wide/wide_opclosure.cpp

```diff
@@ -0,0 +1,191 @@
+// https://github.com/AcademySoftwareFoundation/OpenShadingLanguage
+
+
+#include <iostream>
+#include <cmath>
+
+#include <OSL/genclosure.h>
+#include <OSL/batched_shaderglobals.h>
+
+#include "oslexec_pvt.h"
+
+#include "define_opname_macros.h"
+
+OSL_NAMESPACE_ENTER
+namespace __OSL_WIDE_PVT {
+
+OSL_USING_DATA_WIDTH(__OSL_WIDTH)
+
+
+OSL_BATCHOP void
+__OSL_MASKED_OP(add_closure_closure) (void *bsg_, void *wide_out_, 
+        void *wide_closure_a_, void* wide_closure_b_, unsigned int mask_value)
+{
+    // TODO avoid allot if !a or !b?
+    auto *bsg = reinterpret_cast<BatchedShaderGlobals *>(bsg_);
+    Wide<ClosureColorPtr> wide_out (wide_out_);
+    Wide<ClosureColorPtr> wide_closure_a (wide_closure_a_);
+    Wide<ClosureColorPtr> wide_closure_b (wide_closure_b_);
+    Mask mask (mask_value);
+    
+    ClosureAdd *add = bsg->uniform.context->batched<__OSL_WIDTH>().closure_add_allot ();
+
+    // Currently this is done as AOS, future work may improve this by converting to SOA
+    OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+    for (int lane = 0; lane < __OSL_WIDTH; ++lane) {
+        if (mask[lane]) {
+            ClosureAdd *a = &add[lane];
+            a->id = ClosureColor::ADD;
+            a->closureA = wide_closure_a[lane];
+            a->closureB = wide_closure_b[lane];
+        }
+    }
+    
+    // The input closures may be aliased with the output, so we
+    // do the write-back to the output after we've set-up the
+    // new result
+    OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+    for (int lane = 0; lane < __OSL_WIDTH; ++lane) {
+        if (mask[lane]) {
+            ClosureColorPtr a = (ClosureColorPtr)&add[lane];
+            wide_out[lane] = a;
+        }
+    }
+}
+
+OSL_BATCHOP void
+__OSL_MASKED_OP(mul_closure_color) (void *bsg_, void *wide_out_,  
+        void *wide_closure_a_,  void *wide_weight_, unsigned int mask_value)
+{
+    auto *bsg = reinterpret_cast<BatchedShaderGlobals *>(bsg_);
+    // TODO avoid allot if weight = 0 or closure = null?
+    Wide<ClosureColorPtr> wide_out (wide_out_);
+    Wide<Color3>          wide_weight (wide_weight_);
+    Wide<ClosureColorPtr> wide_closure_a (wide_closure_a_);
+    Mask mask(mask_value);
+
+    ClosureMul *mul = bsg->uniform.context->batched<__OSL_WIDTH>().closure_mul_allot();
+
+    // Currently this is done as AOS, future work may improve this by converting to SOA
+    OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+    for (int lane = 0; lane < __OSL_WIDTH; ++lane) {
+         if (mask[lane]) {
+           ClosureMul *m = &mul[lane];
+           m->id = ClosureColor::MUL;
+           m->weight = wide_weight[lane];
+           m->closure = wide_closure_a[lane];
+        }
+    }
+    
+    // The input closure may be aliased with the output, so we
+    // do the write-back to the output after we've set-up the
+    // new result
+    OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+    for (int lane = 0; lane < __OSL_WIDTH; ++lane) {
+        if (mask[lane]) {
```

Comment:
if `wide_out `is changed to Masked, can remove this mask check allowing calculating of `m` to happen unmasked.